ProcessLib/DensityDrivenFlow/DensityDrivenFlowFEM.h

```diff
@@ -0,0 +1,251 @@
+            auto const beta =
+                _process_data.thermal_expansion_coefficient(t, pos)[0];
+
+            auto const density_fluid = _process_data.density_fluid(t, pos)[0];
+            auto const density_solid = _process_data.density_solid(t, pos)[0];
+            auto const specific_storage =
+                _process_data.specific_storage(t, pos)[0];
+            auto const intrinsic_permeability =
+                _process_data.intrinsic_permeability(t, pos)[0];
+            auto const viscosity0 = _process_data.viscosity(t, pos)[0];
+
+            auto const porosity = _process_data.porosity(t, pos)[0];
+
+            auto const specific_heat_capacity_solid =
+                _process_data.specific_heat_capacity_solid(t, pos)[0];
+            auto const specific_heat_capacity_fluid =
+                _process_data.specific_heat_capacity_fluid(t, pos)[0];
+            double const heat_capacity =
+                density_solid * specific_heat_capacity_solid * (1 - porosity) +
+                density_fluid * specific_heat_capacity_fluid * porosity;
+
+            auto const thermal_conductivity_solid =
+                _process_data.thermal_conductivity_solid(t, pos)[0];
+            auto const thermal_conductivity_fluid =
+                _process_data.thermal_conductivity_fluid(t, pos)[0];
+            double thermal_conductivity =
+                thermal_conductivity_solid * (1 - porosity) +
+                thermal_conductivity_fluid * porosity;
+
+            auto const& sm = _shape_matrices[ip];
+            auto const& wp = _integration_method.getWeightedPoint(ip);
+            auto Ktt = local_K.template block<num_nodes, num_nodes>(0, 0);
+            auto Mtt = local_M.template block<num_nodes, num_nodes>(0, 0);
+            auto Kpp = local_K.template block<num_nodes, num_nodes>(num_nodes,
+                                                                    num_nodes);
+            auto Mpp = local_M.template block<num_nodes, num_nodes>(num_nodes,
+                                                                    num_nodes);
+            auto Bp = local_b.template block<num_nodes, 1>(num_nodes, 0);
+
+            double T_int_pt = 0.0;
+            double p_int_pt = 0.0;
+            // Order matters: First T, then P!
+            NumLib::shapeFunctionInterpolate(local_x, sm.N, T_int_pt, p_int_pt);
+
+            double const delta_T(T_int_pt - temperature0);
+            // TODO include this via material lib
+            double density_water_T = density_fluid * (1 - beta * delta_T);
+            // TODO include viscosity computations via material lib
+            // double const viscosity = viscosity0 * std::exp(- delta_T/75.0);
+            double const perm_visc =
+                intrinsic_permeability / viscosity0;
+
+            auto velocity = (- perm_visc *
+                             (sm.dNdx * p_nodal_values - density_water_T * b))
+                                .eval();
```

Comment:
Could you please give the concrete type for velocity (also const) and drop the eval?